### ISSUE TITLE
Can't split up subjects and can't run all subjects at once. [BUGS]

### ISSUE THREAD

**Issue by wiheto** [nbraingroup/fmridenoise#40]
So I've been trying to get fmridenoise to complete for an entire dataset for over 2 weeks now. Still not managed. And I've run into two problems and I'm stuck how to solve them. 

1. My original plan was to run each subject in parallel on a cluster. This however cannot be done it seems because the group files seem to be overriding each other and the jobs never finish. I'm unsure where the conflict in doing this lies. But this should probably be fixed as, running per subject would increase speed. 
2. After realizing (1) wasn't working, I then specified all of the subjects on one node. However, despite setting the work_dir (the change I just made in a PR that got merged) /tmp/ on the compute node is still getting full. I sshed into the compute node and it is the nifti files which are getting stored in /tmp and I am really unsure why (example of the files being stored):

> ..._space-MNI152NLin2009cAsym_desc-preproc_bold_denoised_pipeline-24HMP_8Phys_SpikeReg_4GS.nii.gz

I've looked at the code. And I can carry on looking, but I cannot see any reason why something should be saved in /tmp/ when -w is set now (most other things are getting stored in the work directory). This sets base_dir and I can't see that ever being overwritten at any point. I am missing it obviously. But perhaps whoever wrote the /tmp/ routines would be able to identify this.  

Finally, and perhaps more importantly The dataset I'm trying to run on is ~28 subjects, 2 sessions and a couple of short runs per task (up to 10). It is not a large dataset by any means. I think it is because there are so many runs that makes fmridenoise take a lot of time (but I don't understand why that would be the case). I am unable to identify why fmridenoise takes this long and uses so much memory in tmp. At the moment I can't run all subjects at once cause of memory issues and /tmp/ issues and I cannot split all the subjects up because there seems to be some type of conflict which causes jobs just to never end.  

(Also, sorry if I sound slightly frustrated, I have been staring at the code for a long time now trying to find the error to fix problem 2 and I just can't see it)

**Comment by wiheto** [comment]
So I think I've figured out what the problem is with (1). It seems to be some problem with multiproc when an error is raised. I don't really have much experience with multiproc so not sure if this is common. 

So when I only run one subject, an error like this is raised in the error log: 
```
Exception: Subnodes of node: QualityMeasures failed:
Subnode 0 failed
Error: Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/utils.py", line 104, in nodelist_runner
    result = node.run(updatehash=updatehash)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 479, in run
    result = self._run_interface(execute=True)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 569, in _run_interface
    return self._run_command(execute)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 662, in _run_command
    result = self._interface.run(cwd=outdir)
  File "/opt/conda/lib/python3.6/site-packages/nipype/interfaces/base/core.py", line 382, in run
    runtime = self._run_interface(runtime)
  File "/opt/conda/lib/python3.6/site-packages/fmridenoise/interfaces/quality_measures.py", line 97, in _run_interface
    corr = pearsonr(group_corr_vec[value[0], i], group_conf_summary['mean_fd'].values[value[0]])
  File "/opt/conda/lib/python3.6/site-packages/scipy/stats/stats.py", line 3392, in pearsonr
    raise ValueError('x and y must have length at least 2.')
ValueError: x and y must have length at least 2.
```

Which I think makes sense that this is raised because there is one subject being run and it can't make a correlation with a single subjects measures.  

But the problem is the jobs are still running and don't return.

```
191025-22:12:46,452 nipype.workflow INFO:
         [MultiProc] Running 5 tasks, and 0 jobs ready. Free memory (GB): 167.46/168.46, Free processors: 19/24.
                     Currently running:
                       * fmridenoise_wf.QualityMeasures
                       * fmridenoise_wf.QualityMeasures
                       * fmridenoise_wf.QualityMeasures
                       * fmridenoise_wf.QualityMeasures
                       * fmridenoise_wf.QualityMeasures
```

It is currently 12.00 the next day and this is the last log entry that exists. 

So this is a little more info about problem (1). But no idea how to solve it!

**Comment by wiheto** [comment]
Alright. So I don't know the solution to nr 1. But the fix in #41 means I can get it to run on the dataset within 4 hours (24 cores). So now the timings are acceptable (compared to 48 nodes with 24 cores not completing after 48 hours, which is the case with option 1). 

I think some documentation may be needed that you can't run per subject and do the same mistakes I did for the last 2 weeks. What I did was the recommended way to run fmriprep, so it is definitely likely that other people do the same mistake I did. 

However, the work directory is: ~400GB in size trying to be stored in /tmp/ (now in the working directory)! The original fmriprep derivatives is 148GB.  I think some ongoing cleanup is needed (and then ). Cleanup_tempdirs is never called. However, and more importantly, line 14 that currently reads: 

> os.makedirs(ret, exist_ok=True) # May be unsafe?

Is definitely unsafe if there is ongoing cleanup.

**Comment by SiegfriedWagner** [comment]
Cleanup is never run (https://github.com/nbraingroup/fmridenoise/issues/15) as it was bugging sometimes. Especially when any exception was raised during execution, and sometimes in multiproc based on my observation. This part of code, responsible for handling temp files, have to 'hardened' and maybe replaced with with solution that is already implemented in nipype (if there is any).

**Comment by wiheto** [comment]
Since I managed to find the problem about specifying the work directory, the temp files not being cleaned up is a known issue, I will open up the problem regarding splitting subjects/warning people not to split subjects in a separate issue so it is a little more focused. Thanks!
